### ISSUE TITLE
Calling external-mutation-free from immutable gives ImpureMethodCall

### ISSUE THREAD

**Issue by vudaltsov** [vimeo/psalm#3065]
https://psalm.dev/r/82396a932a

**Comment by ygottschalk** [comment]
### TL;DR
Expected behavior, not a bug.

### Why you ask?
From the docs:

> @psalm-immutable
Used to annotate a class where every property is treated by consumers as @psalm-readonly and every instance method is treated as @psalm-mutation-free

This means that `LessThan->accept` is treated as `@psalm-mutation-free`. `Visitor->visitLess` is annotated with `@psalm-external-mutation-free`.
Because of latter, there is no guarantee, that `Visitor->visitLess` will not 'mutate' itself. And this guarantee would be needed for the method call in `LessThan->accept`. Putting it in other words: a method annotated with `@psalm-mutation-free` would be allowed to call pure functions and mutation-free methods , but not external-mutation-free methods.
Therefore, this is expected behaviour.

### See also
Really great article explaining 'purity' on it's different levels:
https://psalm.dev/articles/immutability-and-beyond (At the bottom is the purity scale which might help here the most) 

And 2 versions of your script:
https://psalm.dev/r/4d61ac22c3
and a version without any errors
https://psalm.dev/r/c49ebbd750

**Comment by vudaltsov** [comment]
@ygottschalk , thank you very much for the explanations!
